LoRaEngine/LoraKeysManagerFacade/FCntCacheCheck.cs

```diff
@@ -42,41 +43,63 @@ public FCntCacheCheck(ILoRaDeviceCacheStore deviceCache)
+                throw new ArgumentException(errorMsg);
+            }
+
             if (!string.IsNullOrEmpty(abpFcntCacheReset))
             {
-                this.deviceCache.KeyDelete(devEUI);
+                using (var deviceCache = new LoRaDeviceCache(this.deviceCache, devEUI, gatewayId))
+                {
+                    if (await deviceCache.TryToLockAsync())
+                    {
+                        if (deviceCache.TryGetInfo(out var deviceInfo))
+                        {
+                            // only reset the cache if the current valud is larger
+                            // than 1 otherwise we likely reset it from another device
+                            // and continued processing
+                            if (deviceInfo.FCntUp > 1)
+                            {
+                                log.LogDebug("Resetting cache. FCntUp: {fcntup}", deviceInfo.FCntUp);
+                                deviceCache.ClearCache();
+                            }
+                        }
+                    }
+                }
+
                 return (ActionResult)new OkObjectResult(null);
             }
 
             // validate input parameters
             if (!uint.TryParse(fCntDown, out uint clientFCntDown) ||
-                !uint.TryParse(fCntUp, out uint clientFCntUp) ||
                 string.IsNullOrEmpty(gatewayId))
             {
-                string errorMsg = "Missing FCntDown or FCntUp or GatewayId";
+                string errorMsg = "Missing FCntDown or GatewayId";
                 throw new ArgumentException(errorMsg);
             }
 
-            newFCntDown = this.GetNextFCntDown(devEUI, gatewayId, clientFCntUp, clientFCntDown);
+            newFCntDown = await this.GetNextFCntDownAsync(devEUI, gatewayId, clientFCntUp, clientFCntDown);
 
             return (ActionResult)new OkObjectResult(newFCntDown);
         }
 
-        public uint GetNextFCntDown(string devEUI, string gatewayId, uint clientFCntUp, uint clientFCntDown)
+        public async Task<uint> GetNextFCntDownAsync(string devEUI, string gatewayId, uint clientFCntUp, uint clientFCntDown)
         {
             uint newFCntDown = 0;
             using (var deviceCache = new LoRaDeviceCache(this.deviceCache, devEUI, gatewayId))
             {
-                if (deviceCache.TryToLock())
+                if (await deviceCache.TryToLockAsync())
                 {
                     if (deviceCache.TryGetInfo(out DeviceCacheInfo serverStateForDeviceInfo))
                     {
                         newFCntDown = ProcessExistingDeviceInfo(deviceCache, serverStateForDeviceInfo, gatewayId, clientFCntUp, clientFCntDown);
                     }
                     else
                     {
-                        var state = deviceCache.Initialize(clientFCntUp, clientFCntDown + 1);
-                        newFCntDown = state.FCntDown;
+                        newFCntDown = clientFCntDown + 1;
```

Comment:
Is there a reason for this change? The code looks similar to me.

Reply:
Fixed. Thanks.